### PR TITLE
Chat App example

### PR DESCRIPTION
**Pull request by adlrocha**
# Goal

This PR implements a chat application example to showcase the use of smart-records. This example provides the code to start chat clients that connect to a smart-record server, and exchange messages by updating records in the server.
- [x] simple example of chat app using STDIN/STDOUT
- [x] chat app using UI
- [x] Debug tool to read Record from SR server.
- [ ] Documentation

### CODE REVIEW

**Comment by adlrocha** [comment]
@petar, I've been working on this simple chat application to showcase the use of smart-records. Let me know what you think.

File: examples/chat-app/debug/debug.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	// Get Record stored
+	fmt.Println("[*] Getting updated record from peer")
+	out, err := smClient.Get(ctx, *room, info.ID)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println("[*] Record:")
+	for k, v := range *out {
+		fmt.Println("Value for peer: ", k.String(), " - ", v)
```
Comment (petar) [inline examples/chat-app/debug/debug.go:62]: Eventually, make sure the default String() method of syntactic Nodes prints out their JSON representation (which will be our official written representation). For semantics nodes, the String() method should just be Disassemble().String().

File: examples/chat-app/messages.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	"context"
+	"math/big"
+	"sync"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+	"github.com/libp2p/go-smart-record/ir"
+	"github.com/libp2p/go-smart-record/protocol"
+)
+
+type clientConfig struct {
```
Comment (petar) [inline examples/chat-app/messages.go:13]: As we talked earlier, modularize/simplify the implementation by implementing a seprate io.ReadCloser and io.WriteCloser that use an SR client under the hood.

File: examples/chat-app/debug/debug.go (hunk 1 above)
Comment (adlrocha) [reply examples/chat-app/debug/debug.go]: Deferring this to: https://github.com/libp2p/go-smart-record/issues/23

**Comment by adlrocha** [comment]
@petar, I added an additional example removing the UI and using STDIN/STDOUT to read and write messages. I will need to upgrade all the examples once we merge the pending PRs (in order to use metadata, accommodate the separation between semantic and syntactic nodes, etc.).

Also, once this is done I will like to discuss how we want to generate the documentation for this example and the smart-records overall. Thanks!

File: examples/chat-app/main.go [hunk 3]
```diff
@@ -144,21 +166,35 @@ func main() {
...
+		client.syncMessages(outCh)
+
+		// Periodically listen for new messages.
+		msgSyncTicker := time.NewTicker(syncTime)
+		defer msgSyncTicker.Stop()
+		for {
+			select {
+			case <-msgSyncTicker.C:
+				// We periodically fetch the smrat-record to check if there are new messages.
+				client.syncMessages(outCh)
```
Comment (petar) [inline examples/chat-app/main.go:193]: Same comment as in the GC PR: The timer should be reset each time after the completion of syncMessages

**Review by petar — APPROVED** [review]
Just one small comment needs addressing and then it's ready to go. Feel free to merge it after the change. LGTM.

**Comment by petar** [comment]
> @petar, I added an additional example removing the UI and using STDIN/STDOUT to read and write messages. I will need to upgrade all the examples once we merge the pending PRs (in order to use metadata, accommodate the separation between semantic and syntactic nodes, etc.).
> 
> Also, once this is done I will like to discuss how we want to generate the documentation for this example and the smart-records overall. Thanks!

I think, inline comments are sufficient for this example. In addition to the chat app, I want to add one much simpler example which implements a reader/writer channel between two nodes, that uses a smart record as the communication medium. For this example, we may consider writing a tutorial in text. But let's defer these things for now until we are fully migrated to go-routing-language.

**Pull request merged**